File: Algorithm/src/Graph/exam/Coin_Change.java
package Graph.exam;

import java.util.*;

public class Coin_Change {
    public static int coinChange(int[] coins, int amount) {
        if (amount == 0) return 0;

        boolean[] visited = new boolean[amount + 1];
        Queue<int[]> queue = new LinkedList<>();

        // 초기 상태 추가
        queue.offer(new int[]{amount, 0});
        visited[amount] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int remaining = current[0];
            int numCoins = current[1];

            // 금액을 정확히 맞춸을 경우
            if (remaining == 0) {
                return numCoins;
            }

            // 가능한 동전으로 다음 상태 탐색
            for (int coin: coins) {
                int next = remaining - coin;
                if (next >= 0 && !visited[next]) {
                    queue.offer(new int[]{next, numCoins + 1});
                    visited[next] = true;
                }
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] coins = {1, 2, 5};
        int amount = 11;

        System.out.println(coinChange(coins, amount));
    }
}

File: Algorithm_Test_practice/mid/Coin_change.py
from collections import deque

def solution(coins, amount):
    # BFS로 최소 동전수 구하기
    queue = deque()
    visited = [False] * (amount + 1)

    # 시작점 예약 (cur, count)
    queue.append((0, 0))
    visited[0] = True

    # while queue:
    while queue:
        # 현재 노드 방문
        cur, count = queue.popleft()

        # if cur == amount: return count
        if cur == amount:
            return count

        # 다음 노드 방문
        for coin in coins:
            next = cur + coin
            if next <= amount and not visited[next]:
                queue.append((next, count + 1))
                visited[next] = True

    return -1

# 테스트
print(solution([1, 2, 5], 11))
print(solution([2], 3))
print(solution([1], 0))
